File: CÁC BÀI THỰC HÀNH 5/HienThiAnh/app/src/main/java/com/example/displayimage/MainActivity.java
package com.example.displayimage;

import android.annotation.SuppressLint;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.ProgressBar;

import androidx.appcompat.app.AppCompatActivity;

import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;

public class MainActivity extends AppCompatActivity {

    private EditText edtURL;
    private ImageView imageView;
    private ProgressBar progressBar;
    private Button btnload;

    @SuppressLint("MissingInflatedId")
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        edtURL = (EditText) findViewById(R.id.edtURL);
        imageView =(ImageView) findViewById(R.id.imageView);
        progressBar =(ProgressBar) findViewById(R.id.progressBar);
        btnload = (Button) findViewById(R.id.btnLoad);

        btnload.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String url = edtURL.getText().toString();
                new DownloadImageTask().execute(url);
            }
        });
    }

    private class DownloadImageTask extends AsyncTask<String, Integer, Bitmap> {

        @Override
        protected void onPreExecute() {
            super.onPreExecute();
            progressBar.setVisibility(View.VISIBLE);
            imageView.setImageBitmap(null);
        }

        @Override
        protected Bitmap doInBackground(String... urls) {
            String urlString = urls[0];
            Bitmap bitmap = null;
            try {
                URL url = new URL(urlString);
                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
                connection.setDoInput(true);
                connection.connect();
                InputStream input = connection.getInputStream();
                bitmap = BitmapFactory.decodeStream(input);
            } catch (Exception e) {
                e.printStackTrace();
            }
            return bitmap;
        }

        @Override
        protected void onPostExecute(Bitmap result) {
            progressBar.setVisibility(View.GONE);
            if (result != null) {
                imageView.setImageBitmap(result);
            }
        }
    }
}

File: CÁC BÀI THỰC HÀNH 5/UDP/app/src/main/java/com/example/udp/MainActivity.kt
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import com.example.udp.R

class MainActivity : AppCompatActivity() {
    private lateinit var messageInput: EditText
    private lateinit var sendButton: Button
    private lateinit var receivedMessages: TextView
    private val udpPort = 12345
    private val serverIp = "192.168.1.22" 

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_activity)

        messageInput = findViewById(R.id.messageInput)
        sendButton = findViewById(R.id.sendButton)
        receivedMessages = findViewById(R.id.receivedMessages)

        // Bắt đầu lắng nghe tin nhắn
        val receiver = UDPReceiver(udpPort) { message ->
            runOnUiThread { receivedMessages.text = message }
        }
        receiver.start()

        sendButton.setOnClickListener {
            val message = messageInput.text.toString()
            if (message.isNotEmpty()) {
                UDPSender(message, serverIp, udpPort).start()
            }
        }
    }
}
